cuda rpc backward sparse tensor fix

gcramer23 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59609 cuda rpc backward sparse tensor fix**

quick fix for https://github.com/pytorch/pytorch/issues/58755

Differential Revision: [D29335722](https://our.internmc.facebook.com/intern/diff/D29335722)

gcramer23 committed:
cuda rpc backward sparse tensor fix

[ghstack-poisoned]

lw reviewed (approved):
Looks good! I left some style comments which would be nice to address. One more style comment, for the future: it is typical to use commit titles which are actions/orders, like "Do this" or "Make something behave like that". This way by reading through a commit history it looks like reading through a set of instructions (a recipe) for how something was done.

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -953,9 +953,21 @@ std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDa
         value.toPyObjectHolder()->extractTensors();
-    data_ptrs.reserve(tensors.size());
+    auto reserve_size = tensors.size();
     for (const at::Tensor& tensor : tensors) {
-      data_ptrs.emplace_back(tensor.storage().data_ptr());
+      if(tensor.is_sparse()){
+        reserve_size++;
+      }
+    }
```
+1 it will be helpful to add some comments to explain the intention.

gcramer23 replied:
okay added comments to both branches sparse and dense/strided tensor.

mrshenli reviewed (approved):
Thanks a lot for fixing!

gcramer23 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -953,9 +953,21 @@ std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDa
+    }
+    data_ptrs.reserve(reserve_size);
+    for (const at::Tensor& tensor : tensors) {
+      if(tensor.is_sparse()){
+        // sparse tensor is indices and values
+        data_ptrs.emplace_back(tensor.indices().storage().data_ptr());
+        data_ptrs.emplace_back(tensor.values().storage().data_ptr());
+      }else{
+        data_ptrs.emplace_back(tensor.storage().data_ptr());
+      }
```
okay used clang-format.

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -953,9 +953,31 @@ std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDa
+    auto reserve_size = 0;
     for (const at::Tensor& tensor : tensors) {
-      data_ptrs.emplace_back(tensor.storage().data_ptr());
+      if (tensor.is_sparse()) {
+        // Sparse tensor is indices and values. Both are tensors
+        // and contain storage. Therefore reserve_size needs to be
+        // incremented by 2.
+        reserve_size += 2;
+      } else {
+        // A dense/strided tensor contains 1 storage.
+        reserve_size += 1;
+      }
+    }
```
This is much nicer, thanks! The only suggestion I can make is that comments in code are good, but self-explanatory code is even better! (Comments go "stale" quickly, because people forget to update them, whereas this cannot happen to code). Here, for example, if you called the variable `num_storages` then the code itself would already convey most of the meaning that is currently in the comment.

In general I try to only use comments when they're unavoidable, and never to explain _what_ the code is doing: comments should only provide context to explain _why_ some code is doing things in a way that would otherwise look unreasonable.

gcramer23 replied:
I don't think `num_storages` is a good name because the code is storing storage data_ptr in data_ptrs.

gcramer23 replied:
okay Ill update it.

pritamdamania87 reviewed (commented):
Can we have some unit tests that ensure this works as expected?

gcramer23 commented:
@pritamdamania87 

> Can we have some unit tests that ensure this works as expected?

I have been looking into adding a unit test for ivalue::Future::extractDataPtrs. The IValue struct is final, so my understanding is that I can not use gmock to mock it. This complicates things because I have to create a PyObject. I attempted to use pybind to create a python tensor and get the ptr to the PyObject, but it does not work because of linking issues.

I am not familiar with pytorch linking, so that may take some time. Also, If I were to link "Python.h" to the aten directory, is that the correct thing to do?

Not a unit test but I have tested this using the sample program provided in the bug issue https://github.com/pytorch/pytorch/issues/58755.

lw commented:
Our current tests for this logic reside here: https://github.com/pytorch/pytorch/blob/ba09355b12bc4ef0993b9d67e983c9bbbd07bbb5/torch/testing/_internal/distributed/rpc/rpc_test.py#L6056-L6072

They test the Python wrapper for it, just because it was more convenient, but they do work alright. You can add a new method there.

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4736,6 +4762,33 @@ def test_rpc_script_timeout(self):
 
 class TensorPipeAgentRpcTest(RpcAgentTestFixture):
 
+    @skip_if_lt_x_gpu(2)
```
The test below doesn't appear to use GPUs, why is this decorator needed?

gcramer23 replied:
I will remove.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4736,6 +4762,33 @@ def test_rpc_script_timeout(self):
 
+    @skip_if_lt_x_gpu(2)
+    def test_future_extraction_sparse_tensor(self):
+        try:
+            opts = rpc.TensorPipeRpcBackendOptions(init_method=self.init_method)
+            if self.rank == 1:
+                rpc.init_rpc(
+                    "master",
+                    rank=self.rank,
+                    world_size=2,
+                    rpc_backend_options=opts
+                )
+                rref = rpc.remote("server", Server)
+                tensor = build_sparse_tensor()
+                tensor = sparse_tensor_to_rpc_format(tensor)
+                fut = rref.rpc_async().identity(tensor).then(sparse_callback)
+                fut.wait()
+            elif self.rank == 0:
+                rpc.init_rpc(
+                    "server",
+                    rank=self.rank,
+                    world_size=2,
+                    rpc_backend_options=opts
+                )
+        finally:
+            if self.rank < 2:
+                rpc.shutdown()
```
This test won't actually do anything, because DataPtr extraction only occurs for CUDA-enabled Futures, whereas here we're CPU-only.

gcramer23 replied:
My misunderstanding will remove.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4924,6 +4977,34 @@ def forward(self, x, is_rref=False):
 
 class TensorPipeAgentCudaRpcTest(RpcAgentTestFixture):
 
+    @skip_if_lt_x_gpu(2)
+    def test_cuda_future_extraction_sparse_tensor(self):
+        try:
+            opts = rpc.TensorPipeRpcBackendOptions(init_method=self.init_method)
+            if self.rank == 1:
+                opts.set_device_map("server", {self.rank: 0})
+                rpc.init_rpc(
+                    "master",
+                    rank=self.rank,
+                    world_size=2,
+                    rpc_backend_options=opts
+                )
+                rref = rpc.remote("server", Server)
+                tensor = build_sparse_tensor()
+                tensor = sparse_tensor_to_rpc_format(tensor)
+                fut = rref.rpc_async().identity(tensor).then(sparse_callback)
+                fut.wait()
+            elif self.rank == 0:
+                rpc.init_rpc(
+                    "server",
+                    rank=self.rank,
+                    world_size=2,
+                    rpc_backend_options=opts
+                )
+        finally:
+            if self.rank < 2:
+                rpc.shutdown()
```
This test is doing a lot more than strictly necessary: there's no need to spin up the RPC agent and perform network requests just to test the DataPtr extraction logic. If you look at the other tests I pointed at they are much simpler: they create a future and inspect its result to ensure the correct DataPtrs were extracted and used.

gcramer23 replied:
Okay yeah think I had a build problem. I will update it to use the cuda stream.

gcramer23 replied:
I updated `_test_cuda_future_extraction` to add to a tensor on one stream because `fill_` is not implemented for sparse tensors and to have the same test for dense and sparse tensors.

```
NotImplementedError: Could not run 'aten::fill_.Scalar' with arguments from the 'SparseCUDA' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::fill_.Scalar' is only available for these backends: [CPU, CUDA, Meta, QuantizedCPU, QuantizedCUDA, BackendSelect, Named, Conjugate, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, UNKNOWN_TENSOR_TYPE_ID, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode].
```

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -197,6 +197,32 @@ def my_slow_method(self, my_tensor_arg):
 
+class Server:
+    @staticmethod
+    @rpc.functions.async_execution
+    def identity(tensor):
+        fut = torch.futures.Future()
+        fut.set_result(tensor)
+        return fut
```
Why is this using async execution? Couldn't it just return the argument?

gcramer23 replied:
Yes for testing it could just return the argument. This code was removed.

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -953,9 +953,31 @@ std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDa
     std::vector<at::Tensor> tensors =
         value.toPyObjectHolder()->extractTensors();
-    data_ptrs.reserve(tensors.size());
+    auto num_data_ptrs = 0;
```
The argument to `reserve` takes a `size_t` but by using `auto` this will be inferred as `int`, which is "probably" fine but sometimes it could be a smaller type than what we need and, moreover, it's signed whereas `size_t` is unsigned. I think using `size_t` explicitly would be safer.

In general I tend to dislike `auto` because while it's easier to write it makes the code much harder to read. (And, in general, code is read much more often than it's written, which is where the concept of "readability" comes from). I find that `auto` hides the real type of things and could thus sometimes allow an "unexpected" inferred type to slip through unnoticed.

gcramer23 replied:
okay Ill update it.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -6038,37 +6042,70 @@ def test_cuda_future_device_not_cuda(self):
+                if sparse_tensor:
+                    self.assertTrue(torch.eq(tensor.indices(), expected_tensor.indices()).all().item())
+                    self.assertTrue(torch.eq(tensor.values(), expected_tensor.values()).all().item())
+                    self.assertTrue(torch.eq(torch.tensor(tensor.size()), torch.tensor(expected_tensor.size())).all().item())
```
Is there a reason for comparing the sizes by first converting them to a tensor and then doing tensor comparison? I think sizes are tuples hence standard comparison should work:
```suggestion
                    self.assertEqual(tensor.size(), expected_tensor.size())
```
This is much clearer to read and also ends up giving a much more informative error message in case of failure.

gcramer23 replied:
okay Ill update it.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -6038,37 +6042,70 @@ def test_cuda_future_device_not_cuda(self):
+    def _test_cuda_future_extraction(self, wrapper, unwrapper, sparse_tensor):
+        # We check proper CUDA stream synchronization by adding to the tensor
+        # in one stream to get the expected value, and reading it from another stream.
+        if sparse_tensor:
+            tensor = build_sparse_tensor().to("cuda:0")
+            add_tensor = build_sparse_tensor().to("cuda:0")
+            expected_tensor = (tensor + add_tensor).coalesce()
+        else:
+            tensor = torch.zeros((100,), device="cuda:0")
+            add_tensor = torch.ones((100,), device="cuda:0")
+            expected_tensor = tensor + add_tensor
```
This is a preexisting issue, but I'm realizing now that probably we shouldn't create the tensors in the default (unset) stream and then use them in another one without explicitly synchronizing. Perhaps we could in fact move this whole creation step inside `with torch.cuda.stream(stream):` (but before the `torch.cuda._sleep()` call)?

gcramer23 replied:
okay update the code.

lw reviewed (approved):
It looks much better thanks!

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29335722).

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

Differential Revision: [D29335722](https://our.internmc.facebook.com/intern/diff/D29335722)

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29335722).

gcramer23 committed:
Update on "cuda rpc backward sparse tensor fix"


quick fix for https://github.com/pytorch/pytorch/issues/58755

Differential Revision: [D29335722](https://our.internmc.facebook.com/intern/diff/D29335722)

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29335722).
